zendesk/amqp-ts
# Rebuild connection on `consumer cancel` in amqp-ts + Bump amqplib from 0.9.0 to 0.10.3

**Changes**

- Fix to handle empty msg for consumer by rebuilding connection since it means that connection didn't get closed and should be.
- Bump amqplib from 0.9.0 to 0.10.3 - https://github.com/amqp-node/amqplib/compare/v0.9.0...v0.10.3

**Reasons for change**

During AmazonMQ maintenance, queues get transferred from the node in maintenance to one of the 2 remaining nodes available (3 nodes cluster).

When a queue is transferred, it needs to delete the queue.

If consumers are still connected to a queue, it will trigger consumer cancel notification from RabbitMQ server and emits `basic.cancel`.  https://www.rabbitmq.com/consumer-cancel.html

On basic.cancel, amqplib is dispatching an empty message (null). https://github.com/amqp-node/amqplib/blob/main/lib/channel.js#L496

amqp-ts does not know how to deal with msg being set to `null` and will throw an error, since it is caught by a try/catch our application will not crash and consumer will hang forever  -> https://github.com/zendesk/amqp-ts/blob/master/src/amqp-ts.ts#L1031 

Error from sunco-monolith during the maintenance - [Datadog logs](https://zendesk.datadoghq.com/notebook/4772094/amazonmq-pod23-consumers-connection-issues-on-maintenance?fullscreen_end_ts=1676525520000&fullscreen_paused=true&fullscreen_start_ts=1676500380000&fullscreen_widget=lbsul5gk)

**Tests**


- Consumer connected to a queue
- Delete the queue
- RabbitMQ server is emitted basic.cancel to rabbitmq consumers connected to the deleted queue
- worker is handling it gracefully and re-create the queue
Result: OK

- Consumer connected to a queue
- Consumer is processing a long job
- Delete the queue
- Shutdown RabbitMQ for a few mins
- Start RabbitMQ again.

Result: OK, no error and was able to re-create. We might have a risk of duplication issue here. 
--> https://zendesk.slack.com/archives/C04383854TZ/p1677175910137169?thread_ts=1676909575.649359&cid=C04383854TZ

```"error":{"message":"Channel closed","stack":"IllegalOperationError: Channel closed\n    at ConfirmChannel.<anonymous> (/Users/gerard.cao/Code/zendesk/amqp-ts/node_modules/amqplib/lib/channel.js:159:11)\n    at ConfirmChannel.Channel.ack (/Users/gerard.cao/Code/zendesk/amqp-ts/node_modules/amqplib/lib/callback_model.js:233:8)\n    at Message.ack (/Users/gerard.cao/Code/zendesk/amqp-ts/lib/amqp-ts.js:398:27)\n    at Subscriber.ackMessage (/Users/gerard.cao/Code/zendesk/sunco-queue/src/queue/Subscriber.js:129:17)\n    at JobRunner.retryJobLater (/Users/gerard.cao/Code/zendesk/sunco/src/jobs/jobRunner.ts:700:24)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at JobRunner._onError (/Users/gerard.cao/Code/zendesk/sunco/src/jobs/jobRunner.ts:504:17)\n    at /Users/gerard.cao/Code/zendesk/sunco/src/jobs/jobRunner.ts:649:29","stackAtStateChange":"Stack capture: Connection closed: 320 (CONNECTION-FORCED) with message \"CONNECTION_FORCED - broker forced connection closure with reason 'shutdown'
``` 

- [x] Test 3 - Verify if we are loosing messages when queue get deleted and re-created
Result: OK
